File: spring/member-server/src/main/java/com/example/memberserver/member/Member.java
package com.example.memberserver.member;

import com.example.memberserver.common.BaseEntity;
import com.example.memberserver.permission.Permission;
import lombok.Getter;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.ManyToMany;
import java.time.Instant;
import java.util.List;

@Getter
@Entity
public class Member extends BaseEntity {
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Id
    private Long id;

    @Column
    private String name;

    @Column
    private String email;

    @Column
    private String password;

    @ManyToMany
    private List<Permission> permissions;

    public Member() {
        super();
    }

    private Member(Long id, String name, String email, List<Permission> permissions, Instant createdAt, Instant updatedAt) {
        super(createdAt, updatedAt);
        this.id = id;
        this.name = name;
        this.email = email;
        this.permissions = permissions;
    }

    public Member(String name, String email, String password) {
        this.name = name;
        this.email = email;
        this.password = password;
    }

    private Member copy() {
        return new Member(id, name, email, permissions, createdAt, updatedAt);
    }

    public static Member of(MemberDto memberDto) {
        return new Member(memberDto.getName(), memberDto.getEmail(), memberDto.getPassword());
    }

    public Member update(MemberDto memberDto) {
        return null;
    }
}

File: spring/member-server/src/main/kotlin/com/example/memberserver/domain/member/Member.kt
package com.example.memberserver.domain.member

import com.example.memberserver.domain.permission.Permission
import java.time.Instant
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.EntityListeners
import javax.persistence.Id
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType
import javax.persistence.ManyToMany

@Entity
@EntityListeners(AuditingEntityListener::class)
class Member(
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Id
    var id: Long? = null,

    @Column
    val name: String = "",

    @Column
    val email: String,

    @ManyToMany
    var permissions: List<Permission> = listOf(),

    @CreatedDate
    @Column(updatable = false)
    val createdAt: Instant = Instant.EPOCH,

    @LastModifiedDate
    @Column
    val updatedAt: Instant = Instant.EPOCH,

    ) {

    fun assignPermissions(permissions: List<Permission>): Member {
        this.permissions = permissions
        return copy()
    }

    private fun copy(): Member {
        return Member(
            id = id,
            name = name,
            email = email,
            permissions = permissions,
            createdAt = createdAt,
            updatedAt = updatedAt,
        )
    }

    companion object {
        fun of(memberDto: MemberDto): Member {
            return Member(
                name = memberDto.name,
                email = memberDto.email,
            )
        }

    }
}
